### PR TITLE
DEPR: deprecate raise_on_error in .where/.mask in favor of errors=

### LINKED ISSUE #14968: DEPR:change DataFrame.where(..,raise_on_error) -> errors
xref #14878

#14878 deprecated the `raise_on_error` kwarg in favour of the `errors` kwarg 
for `DataFrame.astype()`.

This issue addresses the same deprecation of the `raise_on_error` kwarg for 
`DataFrame.where()`. Again this should be replaced with the `errors` kwarg.

The `errors` kwarg can have the value `raise` or `ignore` with a default 
value of `raise`.

This issue can be assigned to me - @m-charlton

### PR DESCRIPTION
**Pull request by jreback**
closes #14968

### CODE REVIEW

File: pandas/tests/series/test_missing.py [hunk 1]
```diff
@@ -292,15 +292,15 @@ def test_fillna_consistency(self):
...
         result = s.where([True, False],
                          Timestamp('20130101', tz='US/Eastern'),
-                         raise_on_error=False)
+                         errors='ignore')
         assert_series_equal(result, expected)
 
         result = s.where([True, False],
                          Timestamp('20130101', tz='US/Eastern'),
-                         raise_on_error=True)
+                         errors='ignore')
```
Comment (jorisvandenbossche) [inline pandas/tests/series/test_missing.py]: What is the difference between both of the above cases ?

(also , the `errors` has no effect)

**Review by jorisvandenbossche — COMMENTED** [review]
The PR itself looks good, added some minor comments.

But more in general, can you actually give a use case ? I mean a small example where you see the effect of the keyword? 
I haven't been able to make one, and it also seems you didn't have to change any test (the one test you changed, there the keyword has no effect at all). Just wondering, probably I am missing something, but if not I would rather deprecate completely without adding alternative.

**Comment by jreback** [comment]
> But more in general, can you actually give a use case ? I mean a small example where you see the effect of the keyword?
I haven't been able to make one, and it also seems you didn't have to change any test (the one test you changed, there the keyword has no effect at all). Just wondering, probably I am missing something, but if not I would rather deprecate completely without adding alternative.

this really doesn't change anything; just had to fix some actual calls that causes the deprecation warning. The point of this is make things consistent. In the future I actually do want to change the defaults I think (and make 'coerce') possible. This is just conforming things a bit.

File: pandas/tests/series/test_missing.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/series/test_missing.py]: right the kw is pretty much ignored in .where (that's part of the problem); its used in other routines. This is pretty much a mess internally, though making progress.

File: pandas/core/generic.py [hunk 2]
```diff
@@ -5995,24 +6001,44 @@ def _where(self, cond, other=np.nan, inplace=False, axis=None, level=None,
                                            cond_rev="False", name='where',
                                            name_other='mask'))
     def where(self, cond, other=np.nan, inplace=False, axis=None, level=None,
-              try_cast=False, raise_on_error=True):
+              errors='raise', try_cast=False, raise_on_error=None):
+
+        if raise_on_error is not None:
+            warnings.warn(
+                "raise_on_error is deprecated in favor of errors='raise'",
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/generic.py]: oh right

File: pandas/core/computation/expressions.py [hunk 3]
```diff
@@ -200,19 +200,19 @@ def evaluate(op, op_str, a, b, raise_on_error=False, use_numexpr=True,
         op_str: the string version of the op
         a :     left operand
         b :     right operand
-        raise_on_error : pass the error to the higher level if indicated
-                         (default is False), otherwise evaluate the op with and
-                         return the results
+        errors : str, {'raise', 'ignore'}, default 'ignore'
+            'raise' : pass the error to the higher level
+            'ignore' : evaluate the op with and return the results
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/core/computation/expressions.py]: yeah a bunch of copy-paste

**Comment by jorisvandenbossche** [comment]
I know this PR is not changing things, it is just converting the previous situation to a new more consistent one. That is fine. But if it is really a useless keyword, it is also an opportunity to clean things up, and to not just translate the mistakes (?) of the previous keyword to the new one.

Maybe I am missing something, but so my question still is: is there a case `errors='ignore'` actually does something? 
If it currently has no effect, we shouldn't add the new `errors` keyword, but just deprecate the old `raise_on_errors`. Or if adding the new one, have `errors='raise'` as the only option.

**Comment by jreback** [comment]
> Maybe I am missing something, but so my question still is: is there a case errors='ignore' actually does something?
If it currently has no effect, we shouldn't add the new errors keyword, but just deprecate the old raise_on_errors. Or if adding the new one, have errors='raise' as the only option.

We DO need this, it is just not fully working ATM; it was turned off a while back. We need the option to raise, or ignore where the dtype is actually changing (rather than just do it, or in some cases coerce), this is a very common choice which we do in lots of places.

I don't think its a big deal to add the kw. Sure I can prob only allow 'raise' for now.

**Comment by jreback** [comment]
this should be much cleaner now. final thoughts...... @jorisvandenbossche @TomAugspurger

File: pandas/core/internals.py [hunk 4]
```diff
@@ -532,10 +532,20 @@ def astype(self, dtype, copy=False, errors='raise', values=None, **kwargs):
...
+        Coerce to the new type
+
+        dtype : str, dtype convertible
+        copy : boolean, default False
+            copy if indicated
+        errors : str, {'raise', 'ignore'}, default 'ignore'
+            - ``raise`` : allow exceptions to be raised
+            - ``ignore`` : suppress exceptions. On error return original object
+
+        values :
```
Comment (TomAugspurger) [inline pandas/core/internals.py]: Looks like an unfinished docstring here.

**Review by TomAugspurger — COMMENTED** [review]
+1 for doing this now, if it makes things easier in the future.

**Comment by jorisvandenbossche** [comment]
I am fine with having the keyword, I would just update the actual docstring (of the user facing where and mask) to not be misleading (as the current explanation is not correct).

I didn't follow the earlier discussion (https://github.com/pandas-dev/pandas/pull/16821), but just to clarify: what is the idea with `errors` keyword in the future? You say you would like to change this? But to what? 
Because the linked PR actually changed it to not error anymore if a non-matching dtype value is inserted, but to convert to object.

**Comment by jreback** [comment]
> I didn't follow the earlier discussion (#16821), but just to clarify: what is the idea with errors keyword in the future? You say you would like to change this? But to what?
Because the linked PR actually changed it to not error anymore if a non-matching dtype value is inserted, but to convert to object.

well the problem is we are effectively always ``errors='coerce'`` now (even though we don't call it that), and never ``errors='raise```, eg. say you do a ``.where()`` (or lots of other ops like ``.fillna``, ``.setitem`` etc.) on something which would change the dtype, e.g. assign a ``Timestamp`` to an integer array.

Now we just up-convert to ``object``, but providing an option to ``coerce`` or ``raise`` is palable, and in fact you *really* want the default to be ``raise`` (and actually do it rather than up-converting), which is almost always an accident)

**Comment by jreback** [comment]
updated.

**Pull request merged**